lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM.cpp

```diff
@@ -414,19 +343,243 @@ struct ConvertLayoutOpUsingLinearLayoutsConversion
+      std::optional<Value> block =
+          isCrossCTA ? std::make_optional(offsetAndBlock[2]) : std::nullopt;
+      Value offset = offsetAndBlock[0];
+      if (scratchConfig.paddedSize > 0) {
+        assert(llvm::isPowerOf2_32(scratchConfig.paddedStride));
+        assert(llvm::isPowerOf2_32(scratchConfig.paddedSize));
+        auto rshiftVal = llvm::Log2_32(scratchConfig.paddedStride);
+        auto lshiftVal = llvm::Log2_32(scratchConfig.paddedSize);
+        offset = add(shl(lshr(offset, i32_val(rshiftVal)), i32_val(lshiftVal)),
+                     offset);
+      }
+      auto vecAddr = gep(sharedPtrTy, elemTy, smemBase, offset);
+      vecAddr.setInbounds(true);
+      return {vecAddr, block};
+    };
+
+    // register idx -> Value
+    llvm::MapVector<int, Value> outVals;
+    for (int i = 0; i < iterations; i++) {
+      if (i != 0)
+        barrier();
+
+      auto &inRegs = inRegsForIter[i];
+      auto &outRegs = outRegsForIter[i];
+
+      // XXX: CSE applyLinearLayout.
+      for (int j = 0; j < inVals.size() / iterations;
+           j += scratchConfig.inVec) {
+        auto inRegSlice = inRegs[j];
+        auto offsetAndBlock = llvm::to_vector(llvm::make_second_range(
+            applyLinearLayout(loc, rewriter, shmemStoreLayout,
+                              {{kRegister, i32_val(inRegSlice)},
+                               {kLane, laneId},
+                               {kWarp, warpId},
+                               {kBlock, ctaId.value_or(i32_val(0))}})));
+        auto [vecAddr, block] = getVecAddrAndBlock(offsetAndBlock);
+        SmallVector<Value> inValsVec;
+        for (int k = 0; k < scratchConfig.inVec; k++)
+          inValsVec.push_back(inVals[inRegSlice + k]);
+        Value valsVec = packLLVector(loc, inValsVec, rewriter);
+        targetInfo.storeDShared(rewriter, loc, vecAddr, block, valsVec,
+                                /*pred=*/true_val());
+      }
+
+      barrier();
+
+      for (int j = 0; j < outSize / iterations; j += scratchConfig.outVec) {
+        auto outRegSlice = outRegs[j];
+        auto offsetAndBlock = llvm::to_vector(llvm::make_second_range(
+            applyLinearLayout(loc, rewriter, shmemLoadLayout,
+                              {{kRegister, i32_val(outRegSlice)},
+                               {kLane, laneId},
+                               {kWarp, warpId},
+                               {kBlock, ctaId.value_or(i32_val(0))}})));
+        auto [vecAddr, block] = getVecAddrAndBlock(offsetAndBlock);
+        Value valsVec = targetInfo.loadDShared(
+            rewriter, loc, vecAddr, block, vec_ty(elemTy, scratchConfig.outVec),
+            /*pred=*/true_val());
+        for (Value v : unpackLLVector(loc, valsVec, rewriter))
+          outVals[outRegSlice++] = v;
+      }
+    }
+
+    SmallVector<Value> reorderedOutVals;
+    for (size_t i = 0; i < outVals.size(); i++) {
+      reorderedOutVals.push_back(outVals[i]);
+    }
+    return reorderedOutVals;
```

Comment:
Seems like you're *not* reordering the outVals?  Maybe you just want to write

return SmallVector<Value>(outVals.begin(), outVals.end());

?

Also maybe you want to assert that the keys in outVals are 0, 1, 2, ..., N, i.e. they're dense.

Reply:
It is reordering because `outVals` is a MapVector, right?

Reply:
Oooh, right.

Maybe I was confused by the word "reordered" because in a sense it's actually just..."ordered".  Like it has the same "order" as the *keys* in outVals.  Maybe call it outValsVec?